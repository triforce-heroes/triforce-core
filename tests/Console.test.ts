import chalk from "chalk";
import strip from "strip-ansi";
import { describe, expect, it, vitest } from "vitest";

import { fatal } from "@/Console.js";

describe("console", () => {
  chalk.level = 2;

  const samples = [
    ["just message", undefined],
    ["message with details", 123],
    ["message with array-diff", [1, 2, 3], [1, 2, 4]],
    ["message with buffer-diff", Buffer.from("123"), Buffer.from("456")],
    ["message with diff and details", [1], [2], { testA: 1, testB: 2 }],
  ] as const;

  it.each(samples)(
    "function fatal(%j)",
    (
      ...args:
        | [
            message: string,
            expected: unknown,
            received: unknown,
            details?: Record<string, unknown>,
          ]
        | [message: string, details?: unknown]
    ) => {
      let stderrMessage: Uint8Array | string | undefined = undefined;
      let exitCode: number | undefined = undefined;

      vitest.spyOn(process, "exit").mockImplementationOnce((code) => {
        exitCode = Number(code);

        return undefined as never;
      });

      vitest
        .spyOn(process.stderr, "write")
        .mockImplementationOnce((str: Uint8Array | string) => {
          stderrMessage = str;

          return true;
        });

      fatal(...args) as unknown;

      expect(exitCode).toBe(-1);
      expect(stderrMessage).toBeTypeOf("string");
      expect(strip(stderrMessage as unknown as string)).toMatchSnapshot("ansi");
      expect(stderrMessage).toMatchSnapshot("colors");
    },
  );
});
