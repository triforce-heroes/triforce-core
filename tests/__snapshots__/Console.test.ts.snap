// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`console > function fatal("just message") > ansi 1`] = `
"

 ERROR  just message


"
`;

exports[`console > function fatal("just message") > colors 1`] = `
"

 ERROR  just message


"
`;

exports[`console > function fatal("message with array-diff") > ansi 1`] = `
"

 ERROR  message with array-diff

Expected:

[
  1,
  2,
  3
]
Received:

[
  1,
  2,
  4
]
Difference:

 [
   ...
   ...
-  3
+  4
 ]

"
`;

exports[`console > function fatal("message with array-diff") > colors 1`] = `
"

 ERROR  message with array-diff

Expected:

[
  1,
  2,
  3
]
Received:

[
  1,
  2,
  4
]
Difference:

 [
   ...
   ...
[31m-  3[39m
[32m+  4[39m
 ]

"
`;

exports[`console > function fatal("message with buffer-diff") > ansi 1`] = `
"

 ERROR  message with buffer-diff

Difference:

 {
-  0: 49
+  0: 52
-  1: 50
+  1: 53
-  2: 51
+  2: 54
 }

"
`;

exports[`console > function fatal("message with buffer-diff") > colors 1`] = `
"

 ERROR  message with buffer-diff

Difference:

 {
[31m-  0: 49[39m
[32m+  0: 52[39m
[31m-  1: 50[39m
[32m+  1: 53[39m
[31m-  2: 51[39m
[32m+  2: 54[39m
 }

"
`;

exports[`console > function fatal("message with details") > ansi 1`] = `
"

 ERROR  message with details

Details:

123

"
`;

exports[`console > function fatal("message with details") > colors 1`] = `
"

 ERROR  message with details

Details:

123

"
`;

exports[`console > function fatal("message with diff and details") > ansi 1`] = `
"

 ERROR  message with diff and details

Expected:

[
  1
]
Received:

[
  2
]
a:

1
b:

2
Difference:

 [
-  1
+  2
 ]

"
`;

exports[`console > function fatal("message with diff and details") > colors 1`] = `
"

 ERROR  message with diff and details

Expected:

[
  1
]
Received:

[
  2
]
a:

1
b:

2
Difference:

 [
[31m-  1[39m
[32m+  2[39m
 ]

"
`;
